authentication-via-social-networks/README.md

```diff
@@ -0,0 +1,726 @@
+# Authentication via Social Networks
+
+Omniauth schema:
+![alt text](omniauth-schema.png "Omniauth schema")
+
+```ruby
+# Gemfile
+...
+gem "omniauth-facebook"
+gem "omniauth-google-oauth2"
+...
+```
+
+```ruby
+# app/controllers/identities_controller.rb
+class IdentitiesController < ApplicationController
+  before_action :authenticate_user!
+
+  expose(:identities) { current_user.identities }
+  expose(:identity)
+
+  def destroy
+    action = indentity.destroy ? :notice : :alert
+    flash[action] = t "flash.actions.destroy.#{action}", resource_name: Identity.model_name.human
+    redirect_to edit_user_registration_url
+  end
+end
+```
+
+```ruby
+# app/controllers/omniauth_callbacks_controller.rb
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  include OmniauthHelper
+
+  Identity::PROVIDERS.each do |provider|
+    define_method(provider) do
+      show_verification_notice and return unless auth_verified?
```

Comment:
What's the reason to write all this logic in one line? It's so difficult to understand (at least for me) what is going on here.